### PR TITLE
Welcome page optimization & tweaks

### PR DESCRIPTION
**Pull request by connorrothschild**
## Image optimization

Images are preloaded on `/auth` now. For users who auth but do not yet have an account (and are sent to welcome page), this means that images are already loaded and jankiness is prevented. 

The videos below shows onboarding flows for users with "Good 3G" connection before and after preloading:

https://user-images.githubusercontent.com/13339581/149637045-479610c3-24e5-4bcf-aa22-c96221a05e3e.mov

https://user-images.githubusercontent.com/13339581/149637062-302e16d8-9461-4edb-b8e6-3cdbdea3cf6e.mov

The loading of each screen is slow regardless (~11 sec), but when preloading the images are fully loaded as soon as `/welcome` loads, and they do not blink/flicker in

**Note**: This does mean that we're loading images for any user who visits `/auth`, but I don't think this is an issue; the sum of all files is ~1200kb and I don't think (?) loading these images in `useEffect` blocks render

## Vertical screen optimization

Images now have different `x` and `y` positions on vertical screens (screens with a `width/height` aspect ratio of < 1).

Before and after: 

<div style="display: flex;">

</div>

This is also true on super-tall screens/for users who have zoomed out their display:

<div style="display: flex;">

</div>

(Minor but related, we also removed scrollbars on welcome by adding `overflow: hidden`)

## Bonus interactivity

Tbh we can do with or without this change, but I added a lil rotation on hover effect to each individual image. (This meant creating an `<Image />` component for each etc)

It looks like this:

https://user-images.githubusercontent.com/13339581/149637468-e2194772-ea7e-4554-9063-d17cb1922572.mov

I will not be offended if we do not include this lol, was just playing around to learn React and now it exists. 

Feel free to review this a bit more critically because this is my first time ever using React

### CODE REVIEW

**Comment by connorrothschild** [comment]
@Robinnnnn No rush on reviewing this as its the weekend and (like you said) these aren't time-sensitive changes

also not sure why Vercel is failing to deploy... i thought it worked last time but now I get a 404 when I visit the [preview information page](https://vercel.com/gallery-so/gallery/47svF3tysMq1NCsYSeE2Wv4aB5Uj) 🤔

**Comment by kaitoo1** [comment]
@connorrothschild  not sure why you get a 404 for the Vercel deployment page but here's the error I see:

vercel type error

**Comment by connorrothschild** [comment]
Thanks @kaitoo1! Weird I didn't see those type errors in development. Deploy is fixed now, appreciate it!

File: src/scenes/Auth/Auth.tsx [hunk 1]
```diff
@@ -11,12 +11,30 @@ import GalleryRedirect from 'scenes/_Router/GalleryRedirect';
 import breakpoints from 'components/core/breakpoints';
 import Spacer from 'components/core/Spacer/Spacer';
 
+// Preloading images for the welcome screen
+import { animatedImages } from 'src/scenes/WelcomeAnimation/Images';
+
+const preloadImages = () => {
+  animatedImages.forEach((image) => {
+    const img = new Image();
+    img.src = image.src ?? '';
```
Comment (Robinnnnn) [inline src/scenes/Auth/Auth.tsx:20]:
```
animatedImages.forEach((image = '') => {
```

Comment (Robinnnnn) [reply src/scenes/Auth/Auth.tsx:20]: ^ this should remove the need for nullish ??

File: src/scenes/Auth/Auth.tsx [hunk 2]
```diff
@@ -11,12 +11,30 @@ import GalleryRedirect from 'scenes/_Router/GalleryRedirect';
...
   const user = usePossiblyAuthenticatedUser();
   const username = user?.username;
 
+  // Before the welcome screen, we should preload images so that the animation is smooth
+  // NOTE: This means that we preload images for *all* users on /auth, not just new users
+  // This is a necessary tradeoff because the Welcome screen is the first screen for new users,
+  // So there is no 'buffer' screen where we could preload images
+  useEffect(() => {
+    preloadImages();
+  });
```
Comment (Robinnnnn) [inline src/scenes/Auth/Auth.tsx:36]:
```
  useEffect(() => {
    preloadImages();
  }, []);
```

Comment (Robinnnnn) [reply src/scenes/Auth/Auth.tsx:36]:
```js
useEffect(preloadImages, [])
```

File: src/scenes/WelcomeAnimation/Image.tsx [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  const ref = useRef<HTMLDivElement>(null);
+  const [xys, set] = useState([0, 0, 1]);
+  const config = {
+    mass: 1,
+    tension: 170,
+    friction: 75,
+    clamp: false,
+    precision: 0.01,
+    velocity: 0,
+  };
```
Comment (Robinnnnn) [inline src/scenes/WelcomeAnimation/Image.tsx:69]: move this above Image component

File: src/scenes/WelcomeAnimation/Image.tsx [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      ref={ref}
+      onMouseLeave={() => {
+        set([0, 0, 1]);
+      }}
+      onMouseMove={(e) => {
+        if (ref.current) {
+          const rect = ref.current.getBoundingClientRect();
+          set(calc(e.clientX, e.clientY, rect));
+        }
+      }}
```
Comment (Robinnnnn) [inline src/scenes/WelcomeAnimation/Image.tsx:84]:
```js
const handleMouseLeave = useCallback(() => set([0, 0, 1]), []);

const handleMouseMove = useCallback((e) => {
        if (ref.current) {
          const rect = ref.current.getBoundingClientRect();
          set(calc(e.clientX, e.clientY, rect));
        }
      }, [])

...

onMouseLeave={handleMouseLeave}
onMouseMove={handleMouseMove}
```

File: src/scenes/WelcomeAnimation/Image.tsx [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  width,
+  src,
+  fadeInDelay,
+  shouldFadeOut,
+  fadeInGrow,
+  fadeOutGrow,
+  imagesFaded,
+}: Props) {
+  const ref = useRef<HTMLDivElement>(null);
+  const [xys, set] = useState([0, 0, 1]);
```
Comment (Robinnnnn) [inline src/scenes/WelcomeAnimation/Image.tsx:61]: nit: `setXys`

File: src/scenes/WelcomeAnimation/WelcomeAnimation.tsx [hunk 6]
```diff
@@ -154,20 +48,30 @@ export default function WelcomeAnimation({ next }: Props) {
     config: { mass: 10, tension: 550, friction: 140 },
   }));
 
+  // Check for vertical screens and render images in different positions if so
+  const [aspectRatio, setAspectRatio] = useState(0);
+  const windowSize = useWindowSize();
+  useEffect(() => {
+    setAspectRatio(windowSize.width / windowSize.height);
+  }, [windowSize]);
```
Comment (Robinnnnn) [inline src/scenes/WelcomeAnimation/WelcomeAnimation.tsx:56]:
```js
type AspectRatio = 'vertical' | 'horizontal' | undefined

const [aspectRatio, setAspectRatio] = useState<AspectRatio>(undefined);

  useEffect(() => {
    if (windowSize.width / windowSize.height < 1) {
      setAspectRatio('vertical')
      return
    }
    setAspectRatio('horizontal');
  }, [windowSize]);
```

File: src/scenes/WelcomeAnimation/WelcomeAnimation.tsx [hunk 7]
```diff
@@ -256,16 +168,26 @@ const StyledTextContainer = styled.div<{
 
 const StyledMovementWrapper = styled.div<{
   shouldExplode: boolean;
+  aspectRatio: number;
```
Comment (Robinnnnn) [inline src/scenes/WelcomeAnimation/WelcomeAnimation.tsx:171]: aspectRatio: AspectRatio

File: src/scenes/WelcomeAnimation/WelcomeAnimation.tsx [hunk 8]
```diff
@@ -256,16 +168,26 @@ const StyledTextContainer = styled.div<{
...
 }>`
   position: relative;
   transition: transform 1000ms cubic-bezier(0, 0, 0, 1.07);
-  ${({ shouldExplode, animatedImage }) =>
+
+  ${({ shouldExplode, animatedImage, aspectRatio }) =>
     shouldExplode
-      ? `transform: translate(${animatedImage.offsetX}px, ${animatedImage.offsetY}px)`
+      ? `transform: translate(${
+          aspectRatio < 1 && animatedImage.moveOnVertical
```
Comment (Robinnnnn) [inline src/scenes/WelcomeAnimation/WelcomeAnimation.tsx:180]: aspectRatio === 'vertical' ? ...

File: src/scenes/Auth/Auth.tsx (hunk 1 above)
Comment (connorrothschild) [reply src/scenes/Auth/Auth.tsx:20]: `image.src` is what we're checking not `image`, so I think we would need to destructure in the forEach? not totally sure

Comment (Robinnnnn) [reply src/scenes/Auth/Auth.tsx:20]: nope, what you did is totally right! thanks!

**Review by Robinnnnn — APPROVED** [review]

**Review by kaitoo1 — APPROVED** [review]

**Comment by kaitoo1** [comment]
LGTM @connorrothschild !
One small thing, any way we can make the Bonus Interactivity a bit more subtle?
Love the animation but the rotation feels a bit jarring and exaggerated to me - curious to see how it would look if we reduced the amount it tilted by maybe half.

**Comment by kaitoo1** [comment]
And thanks for adding such a detailed description - it made this PR a breeze to review.

**Comment by connorrothschild** [comment]
Good call @kaitoo1 - I agree!

I applied a ceiling of 5deg so that rotation will never exceed that ([see here](https://github.com/gallery-so/gallery/pull/507/files#diff-d88ed163238e95477c86d4a14f30587ace1efe82ac4a0f8ce9a9b5b9049ef74cR43-R58)). Now it looks like this:

https://user-images.githubusercontent.com/13339581/150567479-82bdc54c-0118-4dd5-a8dd-d5bd890e3c8c.mov

A lot more subtle, and we can reduce it more if you see fit.

**Comment by kaitoo1** [comment]
looks great!! thanks for updating 🙏

MERGE ITTTTTTT

**Pull request merged**